File: #2CTinGĐ1619/12CTin/HSG-TPQG/HCM 2015/ATIME.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long 

const long oo = 1e8;
const long maxn = 1020;

int n, m, a, b, k, g, u, v, cc;
int st = 0;
int King[maxn];
int c[maxn][maxn];
int d[maxn];
bool fre[maxn];
vector<vector<int> > node(maxn);

struct data
{
    int u, c;
    data(int uu, int cc)
    {
        this->u = uu;
        this->c = cc;
    }
    bool operator < (const data &rhs) const
    {
        return c > rhs.c;
    }
};
priority_queue<data> heap;

void dijkstra() {
    for (int i = 1; i <= n; i++) d[i] = oo;
    d[a] = k;

    for (int i = 1; i <= n; i++) heap.push(data(i, d[i]));
    memset(fre, true, sizeof(fre));
    while(true) 
    {
        data top = heap.top(); heap.pop();
        if (top.c != d[top.u]) continue;

        int u = top.u;
        fre[u] = false;
        if (u == b) break;
        for (int i = 0; i < node[u].size(); i++)
        {
            int v = node[u][i];
            int tmp;
            if (d[u] >= time[u][v].s and d[u] <= time[u][v].e) tmp = time[u][v].e + 1 - d[u]; else tmp = 0;

            if (fre[v] and d[v] > d[u] + c[u][v] + tmp)
            {
                d[v] = d[u] + c[u][v] + tmp;
                heap.push(data(v, d[v]));
            }
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    freopen("ATIME.inp", "r", stdin);
    freopen("ATIME.out", "w", stdout);

    cin >> n >> m;
    cin >> a >> b >> k >> g;
    for (int i = 0; i < g; i++) cin >> King[i];
    for (int i = 0; i < m; i++) {
        cin >> u >> v >> cc;
        node[u].push_back(v);
        node[v].push_back(u);
        c[u][v] = cc;
        c[v][u] = cc;
    }

    memset(time, 0, sizeof(time));
    for (int i = 1; i < g; i++)
    {
        time[King[i-1]][King[i]].s = st;

        time[King[i-1]][King[i]].e = st + c[King[i-1]][King[i]] - 1;

        time[King[i]][King[i-1]] = time[King[i-1]][King[i]];

        st = time[King[i-1]][King[i]].e + 1;
    }

    dijkstra();
    cout << d[b] - k;
    return 0;
}

File: #2CTinGĐ1619/12CTin/HSG-TPQG/HCM 2015/ATIME.pas
uses math;
type mm = record
        s,e: longint;
     end;
const nmax = 1020;
      oo = 100000000;
var time: array [1..nmax,1..nmax] of mm;
    c: array [1..nmax,1..nmax] of longint;
    d, king: array [1..nmax] of longint;
    n,m,a,b,k,g,u,v,cc,st,i: longint;

procedure dijkstra();
var i,u,s,v,tmp: longint;
    fre: array [1..nmax] of boolean;
begin
        for i:=1 to n do fre[i]:= true;
        for i:=1 to n do d[i]:=oo;
        d[a]:=k;
        while true do begin
                u:= -1; s:=oo;
                for i:=1 to n do
                        if (d[i] < s) and (fre[i]) then begin u:=i; s:=d[i]; end;

                if (u = -1) or (u = b) then exit;
                fre[u]:=false;

                for v:=1 to n do
                        if (c[u][v] > 0) then begin
                                if (d[u] >= time[u,v].s) and (d[u] <= time[u,v].e) then
                                        tmp:=time[u,v].e + 1 - d[u]
                                else
                                        tmp:=0;
                                if (fre[v]) then
                                        d[v]:=min(d[v], d[u] + c[u,v] + tmp);
                        end;
        end;
end;

BEGIN
        assign(input,'ATIME.inp'); reset(input);

        read(n,m,a,b,k,g);
        for i:=1 to g do read(king[i]);
        for i:=1 to m do begin
                read(u,v,cc);
                c[u,v]:= cc;
                c[v,u]:= cc;
        end;
        st:=0;
        for i:=2 to g do begin
                time[king[i-1], king[i]].s:=st;
                time[king[i-1], king[i]].e:= st + c[king[i-1], king[i]] - 1;
                time[king[i], king[i-1]]:= time[king[i-1], king[i]];
                st:= time[king[i-1], king[i]].e + 1;
        end;
        dijkstra();

        assign(output, 'ATIME.out'); rewrite(output); write(d[b] - k); close(output);
END.